tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/localedistance/DistanceTable.java

```diff
@@ -0,0 +1,556 @@
+                node.copySubtablesFrom(wildcardMatch);
+                subtables.put(key.getDesired(), key.getSupported(), node);
+            }
+            return node;
+        }
+
+        /** Copies all subtable mappings from the given node into this one. */
+        private void copySubtablesFrom(Node src) {
+            checkState(subtables.isEmpty());
+            src.subtables.cellSet().forEach(
+                    c -> subtables.put(c.getRowKey(), c.getColumnKey(), new Node(c.getValue().distance)));
+        }
+
+        /**
+         * Writes all the mappings in the distance table sequentially to given Trie in sorted
+         * table order.
+         *
+         * <p>Mappings are written in a top-down recursive visitation with sub-tables inheriting
+         * the current prefix from parent tables via the given Trie span. At each level any
+         * mapped distances are written before recursing into the sub-tables.
+         */
+        private void writeTo(Span trie) {
+            if (distance >= 0 && (distance & DISTANCE_SKIP_SCRIPT) != 0) {
+                // If a node has no distance set, or has been explicitly marked as "skippable",
+                // then write the "default" subtable using the current Trie prefix (effectively
+                // having an "empty" prefix for this case).
+                getAnyNode().writeTo(trie);
+            } else {
+                // In the normal case, just write the mappings explicitly.
+                subtables.rowMap().forEach(
+                        (desired, supportedNodes) -> writeSupported(trie, desired, supportedNodes));
+            }
+        }
+
+        private void writeSupported(Span trie, String desired, Map<String, Node> supportedNodes) {
+            // Collapse any (desired=ANY, supported=ANY) mappings into a single '*' in the trie.
+            if (desired.equals(ANY)) {
+                // If desired is ANY, the only supported subtag must also be ANY.
+                Node node = supportedNodes.get(ANY);
+                checkState(node != null && supportedNodes.size() == 1,
+                        "invalid supported subtags for desired='ANY': %s", supportedNodes);
+                // Remember that ANY != "*", even though it corresponds to "*" in the original
+                // language match rules. Putting "*" in a Trie means something different (but
+                // similar enough to be a bit confusing).
+                trie.with("*", node::writeDistancePlusSubtables);
+            } else {
+                // In the general case, just write the <desired,supported> distance mapping.
+                trie.with(desired, withDesired ->
+                        supportedNodes.forEach((supported, node) -> {
+                            checkState(!supported.equals(ANY),
+                                    "unexpected supported='ANY' subtag: %s", supported);
+                            withDesired.with(supported, node::writeDistancePlusSubtables);
```

Comment:
This is the sort of case the new Trie writing API was designed to make easy to read.

Previously you would have to check lots of code paths to ensure that when a value was to be written, the prefixes had been set. Now, because the extended prefix is passed to the consumer, it's much harder to accidentally misuse this because each new prefix creates a new scope in which it acts.

Compare this with the "toTrie()" method at:
https://github.com/unicode-org/icu/blob/master/icu4j/tools/misc/src/com/ibm/icu/dev/tool/locale/LocaleDistanceBuilder.java#L402

Note also that there were two, almost identical Trie classes in the original code, with slightly different APIs but doing essentially the same thing.